### PR TITLE
[NOD-1127] Implement transaction propagation

### PR DESCRIPTION
**Pull request by someone235**

### CODE REVIEW

File: mempool/mempool.go [hunk 1]
```diff
@@ -1369,6 +1370,43 @@ func (mp *TxPool) HandleNewBlock(block *util.Block, txChan chan NewBlockMsg) err
 	return nil
 }
 
+// HandleNewBlock removes all the transactions in the new block
+// from the mempool and the orphan pool, and it also removes
+// from the mempool transactions that double spend a
+// transaction that is already in the DAG
+func (mp *TxPool) HandleNewBlock(block *util.Block) ([]*util.Tx, error) {
+	// Protect concurrent access.
```
Comment (svarogg) [inline mempool/mempool.go:1378]: Add TODO wherever appropriate to notify about approved transactions in RPC.

File: protocol/blocks.go [hunk 2]
```diff
@@ -0,0 +1,42 @@
...
+	"github.com/kaspanet/kaspad/wire"
+	"sync/atomic"
+)
+
+// OnNewBlock updates the mempool after a new block arrival, and
+// relays newly unorphaned transactions and possibly rebroadcast
+// manually added transactions when not in IBD.
+// TODO(libp2p) Call this function from IBD as well.
+func (m *Manager) OnNewBlock(block *util.Block) error {
+	acceptedTxs, err := m.txPool.HandleNewBlock(block)
```
Comment (svarogg) [inline protocol/blocks.go:16]: Rename somehow that it's clear these are transactions that were accepted to the mempool.

File: protocol/flows/relaytransactions/relaytransactions.go [hunk 3]
```diff
@@ -0,0 +1,227 @@
...
+			return nil
+		}
+	}
+}
+
+func requestInvTransactions(outgoingRoute *router.Route, txPool *mempool.TxPool, dag *blockdag.BlockDAG,
+	sharedRequestedTransactions *SharedRequestedTransactions, inv *wire.MsgTxInv) (requestedIDs []*daghash.TxID,
+	shouldStop bool, err error) {
+
+	idsToRequest := make([]*daghash.TxID, 0, len(inv.TXIDs))
```
Comment (svarogg) [inline protocol/flows/relaytransactions/relaytransactions.go:62]: Why the whole jumping-through-hoops with idsToRequest? I think updateQueues could return it

File: protocol/flows/relaytransactions/relaytransactions.go [hunk 4]
```diff
@@ -0,0 +1,227 @@
...
+	// TODO(libp2p) Consider broadcasting in bulks
+	idsToBroadcast := make([]*daghash.TxID, len(acceptedTxs))
+	for i, tx := range acceptedTxs {
+		idsToBroadcast[i] = tx.Tx.ID()
+	}
+	inv := wire.NewMsgTxInv(idsToBroadcast)
+	return netAdapter.Broadcast(peerpkg.ReadyPeerIDs(), inv)
+}
+
+// readMsgTx returns the next msgBlock in incomingRoute, and populates invsQueue with any inv messages that meanwhile arrive.
```
Comment (svarogg) [inline protocol/flows/relaytransactions/relaytransactions.go:149]: Copy-paste-error: msgBlock -> msgTx

File: protocol/manager.go [hunk 5]
```diff
@@ -0,0 +1,59 @@
...
+	addedTransactions []*util.Tx
+	dag               *blockdag.BlockDAG
+	addressManager    *addrmgr.AddrManager
+
+	transactionsToRebroadcastLock sync.Mutex
+	transactionsToRebroadcast     map[daghash.TxID]*util.Tx
+	lastRebroadcastTime           time.Time
+	sharedRequestedTransactions   *relaytransactions.SharedRequestedTransactions
+
+	isInIBD uint32 // TODO(libp2p) populate this var
```
Comment (svarogg) [inline protocol/manager.go:28]: Please populate now that ibd is merged

File: protocol/transactions.go [hunk 6]
```diff
@@ -0,0 +1,57 @@
...
+	"github.com/kaspanet/kaspad/util/daghash"
+	"github.com/kaspanet/kaspad/wire"
+	"github.com/pkg/errors"
+	"time"
+)
+
+// AddTransaction adds transaction to the mempool and propagates it.
+func (m *Manager) AddTransaction(tx *util.Tx) error {
+	m.transactionsToRebroadcastLock.Lock()
+	defer m.transactionsToRebroadcastLock.Unlock()
```
Comment (svarogg) [inline protocol/transactions.go:15]: Please add empty line after the locking

File: protocol/transactions.go [hunk 7]
```diff
@@ -0,0 +1,57 @@
...
+	acceptedTxs, err := m.txPool.ProcessTransaction(tx, false, 0)
```
Comment (svarogg) [inline protocol/transactions.go:16]: Here too - rename to something that makes it clear these are transactions accepted by mempool

File: protocol/protocol.go [hunk 8]
```diff
@@ -139,38 +123,50 @@ func addPingFlows(router *routerpkg.Router, stopped *uint32, stop chan error, pe
...
+			return ibd.HandleGetBlocks(incomingRoute, outgoingRoute, m.dag)
+		},
+	)
+}
+
+func (m *Manager) addTransactionRelayFlow(router *routerpkg.Router, stopped *uint32, stop chan error) {
+
+	outgoingRoute := router.OutgoingRoute()
+
+	addFlow("RelayedTransactions", router, []wire.MessageCommand{wire.CmdInv, wire.CmdTx}, stopped, stop,
```
Comment (stasatdaglabs) [inline protocol/protocol.go:166]: If we follow convention, this should be "HandleRelayedTransactions".

File: protocol/protocol.go [hunk 9]
```diff
@@ -139,38 +123,50 @@ func addPingFlows(router *routerpkg.Router, stopped *uint32, stop chan error, pe
...
 
 	addFlow("HandleGetBlocks", router, []wire.MessageCommand{wire.CmdGetBlocks}, stopped, stop,
 		func(incomingRoute *routerpkg.Route) error {
-			return ibd.HandleGetBlocks(incomingRoute, outgoingRoute, dag)
+			return ibd.HandleGetBlocks(incomingRoute, outgoingRoute, m.dag)
+		},
+	)
+}
+
+func (m *Manager) addTransactionRelayFlow(router *routerpkg.Router, stopped *uint32, stop chan error) {
```
Comment (stasatdaglabs) [inline protocol/protocol.go:162]: It appears this function is never called.

File: protocol/flows/relaytransactions/shared_requested_transactions.go [hunk 10]
```diff
@@ -0,0 +1,46 @@
...
+	}
+	s.transactions[*txID] = struct{}{}
+	return false
+}
+
+// NewSharedRequestedTransactions returns a new instance of SharedRequestedTransactions.
+func NewSharedRequestedTransactions() *SharedRequestedTransactions {
+	return &SharedRequestedTransactions{
+		transactions: make(map[daghash.TxID]struct{}),
+		Mutex:        sync.Mutex{},
```
Comment (stasatdaglabs) [inline protocol/flows/relaytransactions/shared_requested_transactions.go:44]: This line is redundant.

File: protocol/flows/relaytransactions/relaytransactions.go [hunk 11]
```diff
@@ -0,0 +1,220 @@
...
+	peerpkg "github.com/kaspanet/kaspad/protocol/peer"
+	"github.com/kaspanet/kaspad/protocol/protocolerrors"
+	"github.com/kaspanet/kaspad/util"
+	"github.com/kaspanet/kaspad/util/daghash"
+	"github.com/kaspanet/kaspad/wire"
+	"github.com/pkg/errors"
+	"time"
+)
+
+const timeout = 30 * time.Second
```
Comment (stasatdaglabs) [inline protocol/flows/relaytransactions/relaytransactions.go:17]: Please use DefaultTimeout instead.

File: protocol/flows/relaytransactions/relaytransactions.go [hunk 12]
```diff
@@ -0,0 +1,220 @@
...
+			continue
+		}
+		exists := sharedRequestedTransactions.addIfNotExists(txID)
+		if exists {
+			continue
+		}
+		idsToRequest = append(idsToRequest, txID)
+	}
+
+	msgGetTransactions := wire.NewMsgGetTransactions(idsToRequest)
```
Comment (stasatdaglabs) [inline protocol/flows/relaytransactions/relaytransactions.go:74]: Strictly speaking, you may end up requesting an empty MsgGetTransactions here. Please handle that case.

File: protocol/flows/blockrelay/handle_relay_invs.go [hunk 13]
```diff
@@ -15,14 +13,19 @@ import (
 	mathUtil "github.com/kaspanet/kaspad/util/math"
 	"github.com/kaspanet/kaspad/wire"
 	"github.com/pkg/errors"
+	"time"
 )
 
 const timeout = 30 * time.Second
```
Comment (stasatdaglabs) [inline protocol/flows/blockrelay/handle_relay_invs.go:19]: Please use DefaultTimeout instead.

File: protocol/blocks.go [hunk 14]
```diff
@@ -0,0 +1,43 @@
...
+	"github.com/kaspanet/kaspad/util"
+	"github.com/kaspanet/kaspad/util/daghash"
+	"github.com/kaspanet/kaspad/wire"
+	"sync/atomic"
+)
+
+// OnNewBlock updates the mempool after a new block arrival, and
+// relays newly unorphaned transactions and possibly rebroadcast
+// manually added transactions when not in IBD.
+// TODO(libp2p) Call this function from IBD as well.
```
Comment (stasatdaglabs) [inline protocol/blocks.go:14]: IBD is now implemented. Please call it from there.

File: protocol/flows/ibd/ibd.go [hunk 15]
```diff
@@ -227,6 +235,10 @@ func processIBDBlock(dag *blockdag.BlockDAG, msgIBDBlock *wire.MsgIBDBlock) (sho
 		return false, protocolerrors.Errorf(false, "received delayed block %s "+
 			"during IBD", block.Hash())
 	}
+	err = newBlockHandler(block)
```
Comment (svarogg) [inline protocol/flows/ibd/ibd.go:238]: I'd rather if newBlockHandler was called blockAddedHandler or something like that, newBlockHandler sounds like a constructor

**Pull request merged**